pour importer le resultat dans un fichier excel

manipuler les données et les mettre sous forme de tableau

pour extraire le texte a l interieur du fichier pdf

In [54]:
#!pip install pdfplumber

In [1]:
import pdfplumber

pour les expressions regulieres (regex)

In [2]:
import re

In [3]:
from collections import namedtuple

In [4]:
with pdfplumber.open(r"C:\Users\eree\Desktop\wetransfer_artisanat-zip_2026-01-21_0929\artisanat\artisanat\archive_facture\vente\Facture\maskednew_SOTIC_FAC25-17.pdf") as pdf :
    page=pdf.pages[0]
    text=page.extract_text()

In [5]:
text

'SOUGUI ESHOP Facture\nDate: 29/05/2025\nFacture #: FAC25-17\nClient ID -\nExpéditeur : A envoyer à :\nNom SOUGUI E SHOP Nom SOTIC\nMF 1797689/E Matricule Fiscal 46020C/A/M\n18 Rue Mokhtar Attia Megrine,Ben 152 Avenue 2 Mars GP1 KH 14 , Hammam Lif\nAdresse Adresse\nArous 2033 2050\nTéléphone Téléphone\nRéférence Description PU HT Quantité Prix Total HT\nS1 4 postes 32,00 4,00 1 28,00 TND\nS2 3 postes 32,00 3,00 9 6,00 TND\nMontant Total\nConditions de paiement : 2 24,00 TND\nHT\nTVA 19%\nLes modes de paiement acceptés incluent le chèque et le virement bancaire\nTimbre 1,00\nDétails bancaires\nBanque Agence Compte Clé Total TTC 2 67,560 TND\n00004 00144 1440082831363 3\nDomiciliation : Attijari Bank Tunis\nIBAN TN5904058144008283136345\nArrêté la présente FACTURE à la somme de deux cent soixante et sept dinars cinq\ncent soixante millimes tunisiens\nThank you for your business!\n18 Rue Mokhtar Attia , Megrine, Ben Arous, 2033\nTel: +216 55 466 087 Fax: E-mail: contact@sougui.tn Web: www

## **Données statiques**

In [6]:
text

'SOUGUI ESHOP Facture\nDate: 29/05/2025\nFacture #: FAC25-17\nClient ID -\nExpéditeur : A envoyer à :\nNom SOUGUI E SHOP Nom SOTIC\nMF 1797689/E Matricule Fiscal 46020C/A/M\n18 Rue Mokhtar Attia Megrine,Ben 152 Avenue 2 Mars GP1 KH 14 , Hammam Lif\nAdresse Adresse\nArous 2033 2050\nTéléphone Téléphone\nRéférence Description PU HT Quantité Prix Total HT\nS1 4 postes 32,00 4,00 1 28,00 TND\nS2 3 postes 32,00 3,00 9 6,00 TND\nMontant Total\nConditions de paiement : 2 24,00 TND\nHT\nTVA 19%\nLes modes de paiement acceptés incluent le chèque et le virement bancaire\nTimbre 1,00\nDétails bancaires\nBanque Agence Compte Clé Total TTC 2 67,560 TND\n00004 00144 1440082831363 3\nDomiciliation : Attijari Bank Tunis\nIBAN TN5904058144008283136345\nArrêté la présente FACTURE à la somme de deux cent soixante et sept dinars cinq\ncent soixante millimes tunisiens\nThank you for your business!\n18 Rue Mokhtar Attia , Megrine, Ben Arous, 2033\nTel: +216 55 466 087 Fax: E-mail: contact@sougui.tn Web: www

In [7]:
for line in text.split("\n"):
    print(line)

SOUGUI ESHOP Facture
Date: 29/05/2025
Facture #: FAC25-17
Client ID -
Expéditeur : A envoyer à :
Nom SOUGUI E SHOP Nom SOTIC
MF 1797689/E Matricule Fiscal 46020C/A/M
18 Rue Mokhtar Attia Megrine,Ben 152 Avenue 2 Mars GP1 KH 14 , Hammam Lif
Adresse Adresse
Arous 2033 2050
Téléphone Téléphone
Référence Description PU HT Quantité Prix Total HT
S1 4 postes 32,00 4,00 1 28,00 TND
S2 3 postes 32,00 3,00 9 6,00 TND
Montant Total
Conditions de paiement : 2 24,00 TND
HT
TVA 19%
Les modes de paiement acceptés incluent le chèque et le virement bancaire
Timbre 1,00
Détails bancaires
Banque Agence Compte Clé Total TTC 2 67,560 TND
00004 00144 1440082831363 3
Domiciliation : Attijari Bank Tunis
IBAN TN5904058144008283136345
Arrêté la présente FACTURE à la somme de deux cent soixante et sept dinars cinq
cent soixante millimes tunisiens
Thank you for your business!
18 Rue Mokhtar Attia , Megrine, Ben Arous, 2033
Tel: +216 55 466 087 Fax: E-mail: contact@sougui.tn Web: www.sougui.tn


posotif look behind

## **Date**

In [8]:
date_re = re.compile(r"(?<=Date:\s)\d{2}/\d{2}/\d{4}")
date = date_re.search(text)
print(date)

<re.Match object; span=(27, 37), match='29/05/2025'>


In [9]:
print(date.group())
date = date.group()

29/05/2025


## **Numero de Facture**

In [10]:
numero_Facture_re = re.compile(r"(?<=Facture\s#:\s).+")
numero_Facture = numero_Facture_re.search(text)
numero_Facture = numero_Facture.group()
print(numero_Facture)


FAC25-17


In [11]:
TVA_re = re.compile(r"(?<=TVA\s).+")
TVA = TVA_re.search(text)
TVA = TVA.group()
print(TVA)

19%


## **id_client**

In [12]:
id_Client_re = re.compile(r"(?<=Client ID\s).+")
id_Client = id_Client_re.search(text)
id_Client = id_Client.group()
print(id_Client)


-


## **Nom Client**

In [13]:
nom_client_re = re.compile(r"(?<=Nom SOUGUI E SHOP Nom).+")
nom_client = nom_client_re.search(text)
nom_client =nom_client.group()
print(nom_client)



 SOTIC


## **Matricule Fiscal du Client**

In [14]:
Matricule_client_re = re.compile(r"(?<=Matricule Fiscal ).+")
Matricule_client = Matricule_client_re.search(text)
Matricule_client = Matricule_client.group()
print(Matricule_client)

46020C/A/M


## **Adresse**

In [15]:
import re

lines = text.split("\n")
result = ""

for i, line in enumerate(lines):
    if re.search(r"\bAdresse\s+Adresse\b", line):
        before = lines[i-1] if i > 0 else ""
        current = line
        after = lines[i+1] if i < len(lines) - 1 else ""

        combined = f"{before} {current} {after}"

        # phrases à supprimer (espaces variables)
        patterns = [
            r"18\s*Rue\s*Mokhtar\s*Attia\s*Megrine\s*,?\s*Ben",
            r"Arous\s*2033",
            r"Adresse\s+Adresse"
        ]

        for p in patterns:
            combined = re.sub(p, "", combined, flags=re.IGNORECASE)

        # nettoyage final
        result = re.sub(r"\s{2,}", " ", combined).strip()
        break
adresse_client = result
print(adresse_client)


152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050


In [16]:

print(adresse_client)


152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050


## **Données Dynamiques**

camelot is a library that extracts tables from pdfs

flavor is stream or lattice 


stream is for space seperated tables or for padding margin separated tables 


 lattice is for line seperated tables


 dhaharli mtei line
 ey ey hasb l chat l table hedhi stream
 

camelot sees two tables let s print them

camelot doesn t work for this type of tables

## **Trying to extract the tables with pymupdf**

In [17]:
#!pip install pymupdf

In [18]:
import pymupdf

In [19]:
doc = pymupdf.open(r"C:\Users\eree\Desktop\wetransfer_artisanat-zip_2026-01-21_0929\artisanat\artisanat\archive_facture\vente\Facture\maskednew_SOTIC_FAC25-17.pdf")

In [20]:
page = doc[0]
tables = page.find_tables()


Consider using the pymupdf_layout package for a greatly improved page layout analysis.


In [21]:
tables.tables[0].to_pandas()

,SOUGUI ESHOP,Col1,Col2,Col3,Col4,Col5,Col6,Col7,Col8,Col9,Col10,Col11,Facture,Col13
0,None,None,None,None,None,None,None,,,None,,,None,None
1,None,None,None,None,None,None,None,,,None,,None,None,None
2,,None,None,None,None,None,,,Date:,None,,29/05/2025,None,None
3,,None,None,None,None,None,,,Facture #:,None,,FAC25-17,None,None
4,,None,,,,None,,,Client ID,None,,-,None,None
5,,None,,,,None,,,,None,,,,None
6,,None,,,,None,,,,None,,,,None
7,Expéditeur :,None,None,None,None,None,,,A envoyer à :,None,None,None,None,None
8,Nom,None,SOUGUI E SHOP,None,None,None,,,Nom,None,SOTIC,None,None,None
9,MF,None,1797689/E,None,None,None,,,Matricule Fiscal,None,46020C/A/M,None,None,None


khayba alekher

In [22]:
page.search_for("Description ")

[Rect(171.85000610351562, 250.10499572753906, 223.3479766845703, 260.1400146484375)]

 [x0,y0,x1,y1] l partie l foukaneya mtaa kelmet description ( horizantale) heya y0 donc 250.10499572753906

In [23]:
page.search_for("Téléphone ")

[Rect(27.799999237060547, 209.5049591064453, 70.41500854492188, 219.5399627685547),
 Rect(320.70001220703125, 210.55299377441406, 355.59844970703125, 218.8040008544922)]

y1 : partie loutaneya mtaa kelmet telephone donc                219.5399627685547

In [24]:
page.search_for("Conditions de paiement")



[Rect(36.0, 548.0989990234375, 128.68458557128906, 557.2420043945312)]

In [25]:
conditions_y0 = page.search_for("Conditions de paiement")[0][1]

y0 c est le haut du mot codition de paiement et c est 541.708984375

In [26]:
page = doc[0]
tables = page.find_tables(clip=(0,230,page.rect[2], 550))


tables.tables[0].to_pandas()


,Référence Description PU HT Quantité,Col1,Prix Total HT
0,S1,"4 postes 32,00 4,00","128,00 TND"
1,S2,"3 postes 32,00 3,00","96,00 TND"
2,,,
3,,,
4,,,
5,,,
6,,,
7,,,
8,,,
9,,,


nestaamel l option vertical lines bech nkoul win bedhabt blaset kol colonne

awed tfarej f video l hendi

In [27]:
page = doc[0]
tables = page.find_tables(horizontal_strategy = "text",clip=(0,230,page.rect[2],page.rect[3]))


tables.tables[0].to_pandas()


,Référence Description PU HT Quantité,Col1
0,,None
1,S1,"4 postes 32,00 4,00"
2,,
3,S2,"3 postes 32,00 3,00"


In [28]:
page.search_for("Montant Total")

[Rect(366.1300048828125, 542.574951171875, 421.4800109863281, 552.6099853515625)]

## **trouver les lignes verticales**

trouver les mots clés

In [29]:
page.search_for("Référence")

[Rect(27.799999237060547, 250.70497131347656, 71.19800567626953, 260.739990234375)]

In [30]:
reference_x0 = page.search_for("Référence")[0][0]
print(reference_x0)
reference_x1 = page.search_for("Référence")[0][2]
print(reference_x1)

27.799999237060547
71.19800567626953


In [31]:
page.search_for("Description")

[Rect(171.85000610351562, 250.10499572753906, 220.94497680664062, 260.1400146484375)]

In [32]:

Description_x0 = page.search_for("Description")[0][0]
print(Description_x0)
Description_x1 = page.search_for("Description")[0][2]
print(Description_x1)

171.85000610351562
220.94497680664062


In [33]:
page.search_for("PU HT")

[Rect(324.5, 250.70497131347656, 351.3919677734375, 260.739990234375)]

In [34]:

PU_HT_x0 = page.search_for("PU HT")[0][0]
print(PU_HT_x0)
PU_HT_x1 = page.search_for("PU HT")[0][2]
print(PU_HT_x1)

324.5
351.3919677734375


In [35]:
page.search_for("Quantité")

[Rect(381.9200134277344, 250.70497131347656, 418.1270446777344, 260.739990234375)]

In [36]:

Quantité_x0 = page.search_for("Quantité")[0][0]
print(Quantité_x0)
Quantité_x1 = page.search_for("Quantité")[0][2]
print(Quantité_x1)

381.9200134277344
418.1270446777344


In [37]:
TVA_x0 = page.search_for("TVA")[0][0]
print(Quantité_x0)
TVA_x1 = page.search_for("TVA")[0][2]
print(TVA_x1)

381.9200134277344
383.7249755859375


In [38]:
page.search_for("Prix	Total HT")

[Rect(465.3500061035156, 250.10499572753906, 521.0150756835938, 260.1400146484375)]

In [39]:
page.search_for("TND")

[Rect(524.9560546875, 274.1249694824219, 543.4420776367188, 284.15997314453125),
 Rect(524.9520263671875, 295.1349792480469, 543.4380493164062, 305.16998291015625),
 Rect(523.1547241210938, 622.8809814453125, 543.2349853515625, 633.8079833984375),
 Rect(523.1649169921875, 547.260986328125, 543.2451782226562, 558.18798828125)]

In [40]:
DNT_Montant = page.search_for("TND")[3]
DNT_Montant

Rect(523.1649169921875, 547.260986328125, 543.2451782226562, 558.18798828125)

In [41]:
DNT_Montant_x0 = page.search_for("TND")[0][0]
print(DNT_Montant_x0)
DNT_Montant_x1 = page.search_for("TND")[0][2]
print(DNT_Montant_x1)

524.9560546875
543.4420776367188


la plus proche c est ca Rect(468.7270202636719, 273.3349914550781, 487.2129821777344, 283.3699951171875)

In [42]:
Prix_Total_HT_x0 = page.search_for("Prix	Total HT")[0][0]
print(Quantité_x0)
Prix_Total_HT_x1 = page.search_for("Prix	Total HT")[0][2]
print(Prix_Total_HT_x1)

381.9200134277344
521.0150756835938


## **Les X des Lignes**

In [43]:

#xligne2 =(Description_x0 + reference_x1)/2
xligne2 = reference_x1 +5
#xlignez = Description_x0 -5
#xligne3 =(Description_x1 + PU_HT_x0)/2
xligne3 = PU_HT_x0 -10
#xligne4 = (PU_HT_x1 + Quantité_x0)/2
xligne4 = Quantité_x0
#xligne5 = (Quantité_x1 + Prix_Total_HT_x0 )/2
xligne5 = Quantité_x1 +10

xligne6 = Quantité_x1 +1000
xligne7 = (TVA_x0 + Description_x1)/2
print (xligne2 , xligne3 , xligne4 , xligne5)
print(page.rect[0], page.rect[2])

76.19800567626953 314.5 381.9200134277344 428.1270446777344
0.0 595.2000122070312


In [44]:
vertical_lines = [page.rect[0] ,xligne2 , xligne3 , xligne4 ,xligne5,xligne6,xligne7 , page.rect[2]-5000]


In [45]:
tables = page.find_tables(vertical_lines = vertical_lines ,clip=(0,230,page.rect[2] , conditions_y0 -10) , snap_y_tolerance=7.8)
df =tables[0].to_pandas()
df

,Référence,Description,Col2,Col3,PU HT,Quantité,Col6,Prix Total HT
0,S1,,4 postes,,"32,00","4,0",0,"128,00 TND"
1,S2,,3 postes,,"32,00","3,0",0,"96,00 TND"
2,,,,,,,,
3,,,,,,,,
4,,,,,,,,
5,,,,,,,,
6,,,,,,,,
7,,,,,,,,
8,,,,,,,,
9,,,,,,,,


In [46]:
#mask = df["Description"].isna() | (df["Description"].str.strip() == "")
#df.loc[mask, "Description"] = df.loc[mask, "Col2"]
df["Description"]= df["Col2"]
df = df.drop(columns=["Col2","Col3","Col6"])

In [47]:
df["Numéro Facture"] = numero_Facture
df["id_Client"] = id_Client
df["Date"] = date
df["Client Nom"] = nom_client
df["Matricule Fiscal"] = Matricule_client
df["Adresse"] = adresse_client
df["TVA"] = TVA
df


,Référence,Description,PU HT,Quantité,Prix Total HT,Numéro Facture,id_Client,Date,Client Nom,Matricule Fiscal,Adresse,TVA
0,S1,4 postes,"32,00","4,0","128,00 TND",FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
1,S2,3 postes,"32,00","3,0","96,00 TND",FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
2,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
3,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
4,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
5,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
6,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
7,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
8,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
9,,,,,,FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%


In [48]:
df = df[
    [
        "Référence",
        "Description",
        "PU HT",
        "Quantité",
        "Prix Total HT",
        "Numéro Facture",
        "id_Client",
        "Date",
        "Client Nom",
        "Matricule Fiscal",
        "Adresse",
        "TVA",
    ]
]


In [49]:
df = df[
    df["Référence"].notna() &
    (df["Référence"].str.strip() != "")
]
df


,Référence,Description,PU HT,Quantité,Prix Total HT,Numéro Facture,id_Client,Date,Client Nom,Matricule Fiscal,Adresse,TVA
0,S1,4 postes,"32,00","4,0","128,00 TND",FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
1,S2,3 postes,"32,00","3,0","96,00 TND",FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%


In [50]:
cols = ["Quantité", "PU HT", "Prix Total HT"]
for c in cols:
    df[c] = df[c].fillna("").astype(str).str.strip()

df = df[~(
    (df["Quantité"] == "") &
    (df["PU HT"] == "") &
    (df["Prix Total HT"] ==  "")
)]
df = df.reset_index(drop=True)
df

,Référence,Description,PU HT,Quantité,Prix Total HT,Numéro Facture,id_Client,Date,Client Nom,Matricule Fiscal,Adresse,TVA
0,S1,4 postes,"32,00","4,0","128,00 TND",FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%
1,S2,3 postes,"32,00","3,0","96,00 TND",FAC25-17,-,29/05/2025,SOTIC,46020C/A/M,"152 Avenue 2 Mars GP1 KH 14 , Hammam Lif 2050",19%


In [51]:
df.to_excel(
    r"C:\Users\eree\Desktop\25-17.xlsx",
    index=False
)
